File: topcoder/SRM701D2/ThueMorseGame.java
public class ThueMorseGame {
    public String get(int n, int m) {
        int lastLoseSpot = 0;
        while (lastLoseSpot < n) {
            lastLoseSpot += m + 1;
            while (countSetBits(lastLoseSpot) % 2 == 1)
                ++lastLoseSpot;
        }

        return lastLoseSpot == n ? "Bob" : "Alice";
    }

    private int countSetBits(int n) {
        int result = 0;
        while (n > 0) {
            result++;
            n -= (n & -n);
        }
        return result;
    }
}

File: topcoder/SRM701D2/ThueMorseGame.cpp
#include <cstdio>

int main() {
  int n, m;
  scanf("%d %d", &n, &m);
  
  int lastLoseSpot = 0;
  while (lastLoseSpot < n) {
    lastLoseSpot += (m + 1);
    while (__builtin_popcount(lastLoseSpot) & 1)
      lastLoseSpot++;
  }

  if (lastLoseSpot != n)
    printf("Alice\n");
  else
    printf("Bob\n");
}
